test/e2e/konnect_test.go

```diff
@@ -190,28 +263,73 @@ func createTestRuntimeGroup(ctx context.Context, t *testing.T) string {
 	)
 	require.NoError(t, err)
+	createTimeStamp := time.Now()
+	createTestUUID := uuid.NewString()
+
+	var rgID uuid.UUID
+
+	createRgErr := retry.Do(func() error {
+		rgName := uuid.NewString()
+		createRgResp, err := rgClient.CreateRuntimeGroupWithResponse(ctx, rg.CreateRuntimeGroupRequest{
+			Description: lo.ToPtr(generateTestKonnectRuntimeGroupDescription(t)),
+			Labels: &rg.Labels{
+				"created_in_tests":      "true",
+				"create_test_timestamp": fmt.Sprintf("%d", createTimeStamp.Unix()),
+				"create_test_uuid":      createTestUUID,
+			},
+			Name:        rgName,
+			ClusterType: rg.ClusterTypeKubernetesIngressController,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create runtime group: %w", err)
+		}
+		if createRgResp.StatusCode() != http.StatusCreated {
+			return fmt.Errorf("failed to create RG: code %d, message %s", createRgResp.StatusCode(), string(createRgResp.Body))
+		}
+		if createRgResp.JSON201 == nil || createRgResp.JSON201.Id == nil {
+			return errors.New("No runtime group ID in response")
+		}
+
+		rgID = *createRgResp.JSON201.Id
+		return nil
+	}, retry.Attempts(5), retry.Delay(time.Second))
+
+	require.NoError(t, createRgErr)
 
-	createRgResp, err := rgClient.CreateRuntimeGroupWithResponse(ctx, rg.CreateRuntimeGroupRequest{
-		Description: lo.ToPtr("This is a description"),
-		Labels:      &rg.Labels{"created_in_tests": "true"},
-		Name:        uuid.NewString(),
-		ClusterType: rg.ClusterTypeKubernetesIngressController,
-	})
-	require.NoError(t, err, "failed to create runtime group")
-	require.Equalf(t, http.StatusCreated, createRgResp.StatusCode(), "failed creating RG: %s", string(createRgResp.Body))
-	require.NotNil(t, createRgResp.JSON201)
-	require.NotNil(t, createRgResp.JSON201.Id)
-	id := *createRgResp.JSON201.Id
 	t.Cleanup(func() {
-		_, err := rgClient.DeleteRuntimeGroupWithResponse(ctx, id)
-		assert.NoErrorf(t, err, "failed to cleanup a runtime group: %q", id)
+		// delete the RG used in the test.
+		t.Logf("deleting runtime group %s", rgID)
+		err := retry.Do(
+			func() error {
+				_, err := rgClient.DeleteRuntimeGroupWithResponse(ctx, rgID)
+				if err != nil {
+					t.Logf("failed to delete runtime group %s: %v", rgID.String(), err)
+					return err
+				}
+				return nil
+			},
+			retry.Attempts(5), retry.Delay(time.Second),
+		)
+		assert.NoErrorf(t, err, "failed to delete runtime group %s used in the test", rgID.String())
+
+		// list other RGs also created in the previous code for situations
+		// that RG created, but we timed out on getting the response.
+		// This prevents RG leaks.
+		t.Logf("deleting other runtime groups created in the test")
+		err = retry.Do(
+			deleteRuntimeGroupsCreatedInTest(
+				ctx, *rgClient,
+				t, createTimeStamp, createTestUUID,
+			),
+			retry.Attempts(5), retry.Delay(time.Second),
+		)
+		assert.NoError(t, err, "failed to list and delete other runtime groups created in the test")
```

Comment:
Given we will have the cleanup job taking care of the orphaned runtime groups every 30 minutes, I would suggest dropping the `deleteRuntimeGroupsCreatedInTest` and `create_test_timestamp`, `create_test_uuid` labels added to the runtime group. 

IMO it's not worth taking care of every possible edge case here as anyway we can't be sure that something won't go wrong with deleting. We could assume that the test does not create more than one RG and handle only the "happy path" in the test (APIs are responding with no timeouts which should be the case if we will cleanup runtime groups' roles as in https://github.com/Kong/kubernetes-ingress-controller/pull/4226), leaving the edge cases to be handled by the asynchronous job. If anything goes wrong and we create more than 1 RG, it will get pruned in an hour.

Reply:
Yes, I think it is unnecessary to delete the "leaked" RGs when we have the cleanup job to remove them.